services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/crossregion/S3CrossRegionSyncClientRedirectTest.java

```diff
@@ -0,0 +1,269 @@
+            .isEqualTo("s3." + CROSS_REGION + ".amazonaws.com");
+        verify(mockDelegateClient, times(0)).headBucket(any(HeadBucketRequest.class));
+
+    }
+
+    /**
+     * Call is redirected to actual end point
+     * The redirected call fails because of incorrect parameters passed
+     * This exception should be reported correctly
+     */
+    @Test
+    void apiCallFailure_afterRedirect_isPropagated() {
+
+        when(mockDelegateClient.serviceClientConfiguration()).thenReturn(ENDPOINT_CONFIGURED);
+
+        when(mockDelegateClient.putObject(any(PutObjectRequest.class), any(RequestBody.class)))
+            .thenThrow(redirectException(301, CROSS_REGION))
+            .thenThrow(SERVICE_EXCEPTION);
+        assertThatExceptionOfType(S3Exception.class).isThrownBy(() -> decoratedS3Client.putObject(p -> p.key("Key").bucket(CROSS_REGION_BUCKET), RequestBody.empty()))
+                                                    .withMessage("Invalid id (Service: S3, Status Code: 400, Request ID: 1, "
+                                                                 + "Extended Request ID: A1)");
+    }
+
+    @Test
+    void headBucketCallMade_WhenRedirectMessageDoesNotHaveBucketRegion() throws Exception {
+
+        when(mockDelegateClient.serviceClientConfiguration()).thenReturn(ENDPOINT_CONFIGURED);
+        when(mockDelegateClient.putObject(any(PutObjectRequest.class), any(RequestBody.class)))
+            .thenThrow(redirectException(301, null))
+            .thenReturn(PutObjectResponse.builder().eTag("ABCDEFGH").build());
+
+        when(mockDelegateClient.headBucket(any(HeadBucketRequest.class)))
+            .thenThrow(redirectException(301, CROSS_REGION));
+
+        PutObjectResponse putObjectResponse = decoratedS3Client.putObject(p -> p.key("Key").bucket(CROSS_REGION_BUCKET),
+                                                                          RequestBody.empty());
+        assertThat(putObjectResponse.eTag()).isEqualTo("ABCDEFGH");
+        ArgumentCaptor<PutObjectRequest> requestArgumentCaptor = ArgumentCaptor.forClass(PutObjectRequest.class);
+        verify(mockDelegateClient, times(2)).putObject(requestArgumentCaptor.capture(), any(RequestBody.class));
+
+        assertThat(requestArgumentCaptor.getAllValues().get(0).overrideConfiguration()).isNotPresent();
+        EndpointProvider overriddenEndpointProviderPreCache =
+            requestArgumentCaptor.getAllValues().get(1).overrideConfiguration().get().endpointProvider().get();
+        assertThat(overriddenEndpointProviderPreCache).isInstanceOf(BucketEndpointProvider.class);
+
+        assertThat(((S3EndpointProvider) overriddenEndpointProviderPreCache)
+                       .resolveEndpoint(e -> e.region(Region.US_WEST_2).bucket(CROSS_REGION_BUCKET).build()).get().url().getHost())
+            .isEqualTo("s3." + CROSS_REGION + ".amazonaws.com");
```

Comment:
maybe extract some common validations to a validation method to make the test cases less verbose